README
# ulog
Micro logging library written in C99/C11.

Usage:
void log_to_stderr(
    ulog_level const level,
    char const * const format,
    va_list args
)
{
    vfprintf( stderr, format, args );
    fflush( stderr );
}

ulog_obj const * const ulog = ulog_obj_get();
ulog->op->setup( ulog );
ulog->op->add( ulog, log_to_stderr );
...
UERROR( "my log here" );
...
ulog->op->cleanup( ulog );



ChangeLog
commit 4f22a1e04a1496a4c6751758aa157bf2c4b7a2b1
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sun Jan 31 01:13:45 2016 +0100

    version 2.1

commit 084890c40b6596faa795981c6e3974fb704fb981
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sun Jan 31 01:11:58 2016 +0100

    added new simple test, showing ulog calls from different functions

commit ccb7d61ebf98bb74530c763b71f4f5348da35c69
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sun Jan 31 01:01:50 2016 +0100

    small refactoring of mutex

commit 01dd7ee17142584fa4419e57fa5c229bc8cd5829
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Thu Nov 12 22:54:54 2015 +0100

    updated autotools files to be more robust

commit 650e04654c8c1c501d3cc203f93c1666ba1c7d57
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sun Sep 20 02:16:10 2015 +0200

    Changelog updated

commit 394d5b79b36b0f1f23b2f034194cff5b32243662
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sun Sep 20 02:15:43 2015 +0200

    last fixes; added make check; ulog version bumped to 2.0 in configure

commit 086c4df0547e2a524249c8dd8cae03c22daf4659
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sun Sep 20 00:46:52 2015 +0200

    ulog v2.0

commit 8ebe6b6856f87521c62e70d84ce0aa24482ebfdf
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Fri Sep 4 21:58:46 2015 +0200

    linked list v1.2 ulog_pointer_list -> ulog_listable + ulog_list_ctrl

commit 79aad85869011d0c7e5d33bdcefe58f429842564
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Wed Sep 2 22:55:48 2015 +0200

    missing tab in mutex.c

commit c8a79acb1d5df251931a26888f0ebf4c5235de22
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Wed Sep 2 21:39:46 2015 +0200

    added inline to unlock, fixed some comments

commit 3f9b88a0be0f8c53aa219033eeae3d4471e032e5
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Wed Sep 2 00:42:29 2015 +0200

    word change in status.c comment

commit 452a84d8b8fc653ec70d3b9be354dcc98afb03e1
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Wed Sep 2 00:39:59 2015 +0200

    mutex v 1.2

commit 1bc1f967aa7a50b4c5748b9ada0e2ee370778e9d
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Wed Sep 2 00:37:41 2015 +0200

    fixed tests after mutex default state used guard instead of NULL

commit def0527ef5f6fcf3aed86548412568e16102bf62
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Mon Aug 31 00:07:28 2015 +0200

    small fixes in comments for ulog_mutex

commit 89cde06e1cd8b1db496d9686be4abd9e8a3d5b25
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sun Aug 30 18:59:22 2015 +0200

    ulog_mutex v1.1

commit 5927be9ecb183c7821f283950bb3d11a1d6056c3
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sat Aug 22 17:22:18 2015 +0200

    updated README (C99 instead of C), Changelog

commit 45edeaf95a0f004a1cba805c69aa96081dfbf05c
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sat Aug 22 17:17:33 2015 +0200

    added global verbosity level, meesages below wont reach user handlers

commit 9163518bf78aed228b1d0ce15bc5a05e0f7c136f
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sat Aug 22 16:13:58 2015 +0200

    commented ulog_obj_state

commit 4ea5b93fd36dee2518da9232cfd675b7b6660ae3
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sat Aug 22 16:02:01 2015 +0200

    fixed comment

commit cca5dc3f21239d3aa6cd73605a7838bfec221065
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sat Aug 22 15:55:25 2015 +0200

    added test for ulog_level_to_char()

commit a822fea4c56163e1589b5c7ace6bc00aeb4c11f2
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sat Aug 22 15:52:57 2015 +0200

    removed unused main arguments - main(void) is legal

commit 1cf4fb08f128bd5c25c4b9288c083bb6c0ea3dff
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sat Aug 22 15:45:27 2015 +0200

    added test showing clean compilation with empty ulog call

commit 0d4beaf597277506cc842a6a641cadf16315fea2
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Sat Aug 22 15:44:42 2015 +0200

    added showing log levels in output

commit 35419d0b5556e47d3b9bc892902f6b26bdf209d6
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Thu Jul 16 21:53:23 2015 +0200

    unused include removed; updated old comment

commit b77620a8df59a33ce04003a482034dd17666bfbe
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Mon May 4 10:32:38 2015 +0200

    stylistic fixes

commit d0b0cfd573dbe3e611f8d2cec881ac81907436bd
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Mon May 4 10:18:16 2015 +0200

    now asserting ulog_setup()

commit 72897b84fba60484371510a772d755e470cf2c40
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Mon May 4 10:09:06 2015 +0200

    more tests

commit 4880c092ec4044e9326b862a8e0a9c21505ec84f
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Tue Apr 28 16:12:21 2015 -0700

    global ulog_obj returned by getter is now const (note: state is a const pointer to writable object)

commit 9679a74437cc2af3b344307d657c21bc5d837715
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Tue Apr 28 15:50:41 2015 -0700

    updated changelog

commit 78b1ea8d37f465ea05dfaf6ebbb736977004fb9f
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Tue Apr 28 15:50:21 2015 -0700

    added -Wall -Wextra -pedantic flags to compilation

commit e22ab17bd292f617a39ab22587ad68171bf9877a
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Tue Apr 28 15:49:55 2015 -0700

    getter for global ulog_obj var for better control

commit c2c9404822ae23981783f3d29d278837b95f8788
Author: Mateusz Jemielity <matthew.jemielity@gmail.com>
Date:   Tue Apr 28 08:23:57 2015 -0700


src/mutex.c
/**
 * \author      Mateusz Jemielity matthew.jemielity@gmail.com
 * \brief       C99/C11 implementation of simple mutex wrapper.
 * \date        2015/08/29 17:16:23 AM
 * \file        mutex.c
 * \version     1.2
 *
 *
 **/

#include <ulog/mutex.h>
#include <ulog/status.h> /* ulog_status, ulog_status_descriptive */
#include <ulog/universal.h> /* THREADUNSAFE, UNUSED */

#include <errno.h> /* EALREADY, EBUSY, EINVAL, EIO, ENOMEM */
#include <stdbool.h> /* bool */
#include <stddef.h> /* NULL */
#include <stdlib.h> /* free, malloc */
#if __STDC_NO_THREADS__
# include <pthread.h>
#else /* !__STDC_NO_THREADS__ */
# include <threads.h>
#endif /* __STDC_NO_THREADS__ */

#if __STDC_NO_THREADS__
# define ULOG_MUTEX_SUCCESS 0
#else /* !__STDC_NO_THREADS__ */
# define ULOG_MUTEX_SUCCESS thrd_success
#endif /* __STDC_NO_THREADS__ */

typedef
#if __STDC_NO_THREADS__
    pthread_mutex_t
#else /* !__STDC_NO_THREADS__ */
    mtx_t
#endif /* __STDC_NO_THREADS__ */
mutex_obj;

struct ulog_mutex_state_struct
{
    mutex_obj mutex;
    ulog_mutex_op_table const * op;
};

typedef
#if __STDC_NO_THREADS__
    pthread_mutexattr_t const *
#else /* !__STDC_NO_THREADS__ */
    int
#endif /* __STDC_NO_THREADS__ */
internal_mutex_ctrl_arg;

typedef int
( * internal_mutex_ctrl )(
    mutex_obj * mutex,
    internal_mutex_ctrl_arg arg
);

static internal_mutex_ctrl const ctrl_init_operation =
#if __STDC_NO_THREADS__
    pthread_mutex_init
#else /* !__STDC_NO_THREADS__ */
    mtx_init
#endif /* __STDC_NO_THREADS__ */
;
static internal_mutex_ctrl_arg const default_init_arg =
#if __STDC_NO_THREADS__
    NULL
#else /* !__STDC_NO_THREADS__ */
    mtx_plain
#endif /* __STDC_NO_THREADS__ */
;

typedef int
( * internal_mutex_op )( mutex_obj * mutex );

static inline int
ulog_mutex_init( mutex_obj * mutex )
{
    return ctrl_init_operation( mutex, default_init_arg );
}

#if !__STDC_NO_THREADS__
static inline int
ulog_mutex_destroy( mutex_obj * mutex )
{
    mtx_destroy( mutex );
    return ULOG_MUTEX_SUCCESS;
}
#endif /* !__STDC_NO_THREADS__ */

typedef struct
{
    internal_mutex_op const op;
    char const * const success;
    char const * const failure;
}
generic_operation_arg;

static generic_operation_arg const generic_arg[ 4U ] =
{
    {
        .op = ulog_mutex_init,
        .success = "mutex set up successfully",
        .failure = "failure setting up the underlying mutex mechanism"
    },
    {
        .op =
#if __STDC_NO_THREADS__
            pthread_mutex_destroy
#else /* !__STDC_NO_THREADS__ */
            ulog_mutex_destroy
#endif /* __STDC_NO_THREADS__ */
        ,
        .success = "mutex cleaned up successfully",
        .failure = "failure cleaning up the underlying mutex mechanism"
    },
    {
        .op =
#if __STDC_NO_THREADS__
          pthread_mutex_lock
#else /* !__STDC_NO_THREADS__ */
          mtx_lock
#endif /* __STDC_NO_THREADS__ */
        ,
        .success = "mutex locked successfully",
        .failure = "error locking the underlying mutex mechanism"
    },
    {
        .op =
#if __STDC_NO_THREADS__
          pthread_mutex_unlock
#else /* !__STDC_NO_THREADS__ */
          mtx_unlock
#endif /* __STDC_NO_THREADS__ */
        ,
        .success = "mutex unlocked successfully",
        .failure = "error unlocking the underlying mutex mechanism"
    }
};

typedef enum
{
  SETUP = 0U,
  CLEANUP = 1U,
  LOCK = 2U,
  UNLOCK = 3U
}
generic_arg_type;

static inline bool
valid( ulog_mutex const * const self );

static inline ulog_status
generic_invalid( ulog_mutex const * const self )
{
    UNUSED( self );
    return ulog_status_descriptive( EINVAL, "invalid mutex object" );
}

static inline ulog_status
generic_uninitialized( ulog_mutex const * const self )
{
    UNUSED( self );
    return ulog_status_descriptive( EINVAL, "mutex object uninitialized" );
}

static inline ulog_status
generic_already( ulog_mutex const * const self, char const * const message )
{
    UNUSED( self );
    return ulog_status_descriptive( EALREADY, message );
}

static inline ulog_status
setup_already( ulog_mutex * const self )
{
    return generic_already( self, "mutex already set up" );
}

static inline ulog_status
cleanup_already( ulog_mutex * const self )
{
    return generic_already( self, "mutex already cleaned up" );
}

static inline ulog_status
generic_operation(
    ulog_mutex const * const self,
    generic_operation_arg const * const arg
)
{
    switch( arg->op( &( self->state->mutex )))
    {
        case ULOG_MUTEX_SUCCESS:
            return ulog_status_descriptive( 0, arg->success );
        default:
            return ulog_status_descriptive( EIO, arg->failure );
    }
}

static THREADUNSAFE ulog_status
setup_safe( ulog_mutex * const self );
static THREADUNSAFE ulog_status
cleanup_safe( ulog_mutex * const self );
static inline ulog_status
lock_safe( ulog_mutex const * const self );
static inline ulog_status
unlock_safe( ulog_mutex const * const self );

static ulog_mutex_op_table const default_op =
{
    .setup = setup_safe,
    .cleanup = cleanup_already,
    .lock = generic_uninitialized,
    .unlock = generic_uninitialized
};
static ulog_mutex_op_table const setup_op =
{
    .setup = setup_already,
    .cleanup = cleanup_safe,
    .lock = lock_safe,
    .unlock = unlock_safe
};
/*
 * it would be nice to have an additional "locked" state
 * however this would at least require atomic op changes
 * additionally trying to lock a locked mutex would fail
 * instead of blocking the calling thread
 */

static ulog_mutex_state guard = { .op = &default_op };

static THREADUNSAFE ulog_status
setup_safe( ulog_mutex * const self )
{
    ulog_mutex_state * const state = malloc( sizeof( ulog_mutex_state ));
    if( NULL == state )
    {
        return ulog_status_descriptive(
            ENOMEM,
            "cannot allocate memory for mutex state"
        );
    }
    self->state = state;
    self->state->op = &setup_op;

    ulog_status const result =
        generic_operation( self, &( generic_arg[ SETUP ] ));
    if( !ulog_status_success( result ))
    {
        free( self->state );
        self->state = &guard;
    }
    return result;
}

static THREADUNSAFE ulog_status
cleanup_safe( ulog_mutex * const self )
{
    /*
     * In case of pthreads (and possibly C11 threads, since they seem to use
     * pthreads) the implementation should allow safe destruction of mutex
     * right after it has been unlocked.
     */
    self->op->lock( self );
    self->op->unlock( self );
    ulog_status const result =
        generic_operation( self, &( generic_arg[ CLEANUP ] ));
    if( ulog_status_success( result ))
    {
        free( self->state );
        self->state = &guard;
    }
    return result;
}

static inline ulog_status
lock_safe( ulog_mutex const * const self )
{
    return generic_operation( self, &( generic_arg[ LOCK ] ));
}

static inline ulog_status
unlock_safe( ulog_mutex const * const self )
{
    return generic_operation( self, &( generic_arg[ UNLOCK ] ));
}

static inline THREADUNSAFE ulog_status
setup( ulog_mutex * const self )
{
    if( !valid( self )) { return generic_invalid( self ); }
    return self->state->op->setup( self );
}

static inline THREADUNSAFE ulog_status
cleanup( ulog_mutex * const self )
{
    if( !valid( self )) { return generic_invalid( self ); }
    return self->state->op->cleanup( self );
}

static inline ulog_status
lock( ulog_mutex const * const self )
{
    if( !valid( self )) { return generic_invalid( self ); }
    return self->state->op->lock( self );
}

static inline ulog_status
unlock( ulog_mutex const * const self )
{
    if( !valid( self )) { return generic_invalid( self ); }
    return self->state->op->unlock( self );
}

static ulog_mutex_op_table const op =
{
    .setup = setup,
    .cleanup = cleanup,
    .lock = lock,
    .unlock = unlock
};

static inline bool
valid( ulog_mutex const * const self )
{
    return
        (
            ( NULL != self )
            && (
                (
                    ( &guard == self->state )
                    && ( &default_op == self->state->op )
                )
                || (
                    ( NULL != self->state )
                    && ( &guard != self->state )
                    && ( &setup_op == self->state->op )
                )
            )
            && ( &op == self->op )
        );
}

ulog_mutex
ulog_mutex_get( void )
{
    return ( ulog_mutex ) { .state = &guard, .op = &op };
}

